lib/utils/batcher/batcher.go

```diff
@@ -0,0 +1,172 @@
+)
+
+// Run processes events from the given channel in batches, calling fn for each batch.
+// Events are collected into batches during the time window with maximum threshold number of events.
+//
+// It is useful in cases of high event volume where events can be aggregated and
+// processed together in a single operation. For example, when reacting to role
+// change events and check what users have been affected by this change.
+// Instead of fetching all the users for each role change event
+// you can aggregate roles change events and fetch and process all users only once
+// instead of fetching all the users from each role change event.
+//
+// Example:
+//
+//	   opts := []batcher.Option{
+//	       batcher.WithClock(u.clock),
+//	       batcher.WithWindow(time.Second * 10),
+//	       batcher.WithThreshold(100),
+//	   }
+//
+//		err := batcher.Run(ctx, eventChan, func(batch []Event) error {
+//			log.Printf("Processing %d events", len(batch))
+//			return processEvents(batch)
+//		}, opts...))
+//
+// The function returns when the context is canceled, returning ctx.Err(),
+// or when fn returns an error, returning that error.
+func Run[T any](ctx context.Context, events <-chan T, fn func(batch []T) error, opts ...Option) error {
+	collector := New[T](opts...)
+	return trace.Wrap(collector.Run(ctx, events, fn))
+}
+
+// Option configures a Collector using the functional options pattern.
+type Option func(*Collector[any])
+
+// WithWindow sets the collection window duration.
+func WithWindow(d time.Duration) Option {
+	return func(c *Collector[any]) {
+		if d > 0 {
+			c.window = d
+		}
+	}
+}
+
+// WithThreshold sets the maximum number of events per batch.
+// Default is unlimited (bounded only by window duration).
+func WithThreshold(n int) Option {
+	return func(c *Collector[any]) {
+		if n > 0 {
+			c.threshold = n
+		}
+	}
+}
+
+// WithClock sets a custom clock for testing.
+// Default uses the system clock.
+func WithClock(clock clockwork.Clock) Option {
+	return func(c *Collector[any]) {
+		if clock != nil {
+			c.clock = clock
+		}
+	}
+}
+
+// New creates a new Collector with the given options.
+func New[T any](opts ...Option) *Collector[T] {
+	c := &Collector[T]{
+		window:    10 * time.Second,
+		threshold: 50,
+		clock:     clockwork.NewRealClock(),
+	}
+	// Apply options (type conversion needed for generics)
+	for _, opt := range opts {
+		opt((*Collector[any])(c))
+	}
+	return c
+}
+
+// Collector batches events into time windows.
+type Collector[T any] struct {
+	window    time.Duration
+	threshold int
+	clock     clockwork.Clock
+}
+
+// Run starts collecting events and calls fn for each batch.
+// It blocks until ctx is canceled or an error occurs.
+func (c *Collector[T]) Run(ctx context.Context, events <-chan T, fn func(batch []T) error) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		batch, err := c.CollectBatch(ctx, events)
+		if err != nil {
+			return err
+		}
```

Comment:
Should we return the context error only after processing the last chunk? Keeping the invariant that events are either in the channel or they're pulled from the channel and processed might be nice.

Reply:
I have added the desc 61fea393ca03aa872b3b5af06aad4236b7f61763


The flow depends on the use case. My main use of batcher is processing long running jobs where the ctx is the process context. 

If the will continue to process on canceled context  the handler will return errors producing log noise. 

But still this behavior is possible using raw ` CollectBatch(ctx context.Context, events <-chan T) ([]T, error)` 


Where the common use of batcher that I was thinking was processing events with teleport process context.